diff --git a/IDstage.v b/IDstage.v
index 8d2aa6d..4781021 100644
--- a/IDstage.v
+++ b/IDstage.v
@@ -120,7 +120,18 @@ wire inst_div_wu;
 wire inst_mod_w;
 wire inst_mod_wu;
 
+wire inst_sltui;
+wire inst_slti;
+wire inst_andi;
+wire inst_ori;
+wire xori;
+wire inst_sll_w;
+wire inst_srl_w;
+wire inst_sra_w;
+wire inst_pcaddu12i;
+
 wire need_ui5;
+wire need_ui12;
 wire need_si12;
 wire need_si16;
 wire need_si20;
@@ -314,7 +325,7 @@ regfile u_regfile(
 //assign rkd_value = rf_rdata2;
 
 assign rj_value = 
-    (exe_rf_we && exe_dest == rf_raddr1)? alu_result :////
+    (exe_rf_we && exe_dest == rf_raddr1)? alu_result :
     (mem_rf_we && mem_dest == rf_raddr1)? final_result :
     (rf_we  && dest_reg   == rf_raddr1)? rf_wdata   :
     rf_rdata1;
diff --git a/alu.v b/alu.v
index 0c931f4..a435867 100644
--- a/alu.v
+++ b/alu.v
@@ -20,6 +20,7 @@ wire op_sll;   //logic left shift
 wire op_srl;   //logic right shift
 wire op_sra;   //arithmetic right shift
 wire op_lui;   //Load Upper Immediate
+
 wire op_mul;   //multiply
 wire op_mulh;  //multiply and store high bits
 wire op_mulhu; //unsigned multiply and store high bits
@@ -74,7 +75,7 @@ mul u_mul(
   .x(mul_src1),
   .y(mul_src2),
   .z(mul_result)
-)
+);
 
 
 // 32-bit adder
@@ -157,6 +158,7 @@ always @(posedge clk) begin
   end else begin
     div_data_valid <= 1'b0;
   end
+end
 
   IP_DIV div(
   .aclk                   (clk),
